### PR TITLE
TST: Add unit tests for older timezone issues

### LINKED ISSUE #7825: BUG: Timestamp.replace() behaves naively at DST boundaries (plus bonus segfault!)
``` python

In [1]: import pandas as pd

In [2]: t = pd.Timestamp('2013-11-3', tz='America/Chicago')

In [3]: t
Out[3]: Timestamp('2013-11-03 00:00:00-0500', tz='America/Chicago')

In [4]: t.replace(hour=3)
Out[4]: Timestamp('2013-11-03 03:00:00-0500', tz='America/Chicago')
# This time doesn't exist.

In [5]: pd.Timestamp('2013-11-3 03:00:00', tz='America/Chicago')
Out[5]: Timestamp('2013-11-03 03:00:00-0600', tz='America/Chicago')
```

In trying to replicate the definition of `Timestamp.replace` in [tslib.pyx](https://github.com/pydata/pandas/blob/320f9d4c44d60cc4ae5e08ab4a26dc49d3fd71c1/pandas/tslib.pyx#L421), I encountered this further issue:

``` python

In [6]: from datetime import datetime

In [7]: datetime.replace(t, hour=3)
Segmentation fault: 11
```

Trying another tack, I found this:

``` python

In [1]: from datetime import datetime

In [2]: import pytz

In [3]: t = pytz.timezone('America/Chicago').localize(datetime(2013, 11, 3))

In [4]: t.strftime('%Y-%m-%d %H:%M:%S %Z%z')
Out[4]: '2013-11-03 00:00:00 CDT-0500'

In [5]: t.replace(hour=3).strftime('%Y-%m-%d %H:%M:%S %Z%z')
Out[5]: '2013-11-03 03:00:00 CDT-0500'
# This time still never happened
```

In summary: `datetime` is horribly broken, and the current workaround doesn't actually get all the way around.  I know breaking backward compatibility is painful, and throwing the stdlib out the window is even more so, but this situation reminds me of nothing so forcefully as Microsoft Excel's 1900/1904 issue.

Version information:

``` python

In [1]: from pandas.util.print_versions import show_versions

In [2]: show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.5.final.0
python-bits: 64
OS: Darwin
OS-release: 13.3.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.14.1
nose: 1.3.0
Cython: None
numpy: 1.8.1
scipy: 0.12.1
statsmodels: None
IPython: 1.1.0
sphinx: None
patsy: None
scikits.timeseries: None
dateutil: 2.2
pytz: 2014.4
bottleneck: None
tables: None
numexpr: None
matplotlib: 1.3.0
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
rpy2: None
sqlalchemy: None
pymysql: None
psycopg2: None
```

### LINKED ISSUE #8306: ERR: raise in tz incompat with DatetimeIndex ops
xref #6552

```
idx = pd.date_range(..., tz='UTC')
idx2 = idx.reindex(pd.date_range(..., tz='US/Pacific')
```
- [ ] subtraction of datetime like 
- [ ] intersection / union / difference
- [ ] reindex

### LINKED ISSUE #8531: bug in tz_localize when summer/winter time change
Works with datetime.datetime, does not work with pandas.tslib.Timestamp.
Encounter the issue when trying to resample hourly a series.

```
import pandas as pd
from datetime import datetime as dt
import pytz


tz = 'Europe/Berlin'
d = dt(2014,10,26,02,00)
tz = pytz.timezone(tz)
tz.localize(d)

x = pd.Series(0., [d]).index[0]
x.tz_localize(tz)
```

### LINKED ISSUE #8616: BUG: DatetimeIndex.shift(freq=...) raises near DST boundary
xref #5694 #8531 (?)
xref #8817 

This is presumably caused by the fact that pytz time zones internalize the offset of the current time.

``` python
In [1]: import pandas as pd

In [2]: idx = pd.date_range('2013-11-03', tz='America/Chicago',
   ...:                     periods=6, freq='H')

In [3]: pd.Series(index=idx)
Out[3]: 
2013-11-03 00:00:00-05:00   NaN
2013-11-03 01:00:00-05:00   NaN
2013-11-03 01:00:00-06:00   NaN
2013-11-03 02:00:00-06:00   NaN
2013-11-03 03:00:00-06:00   NaN
2013-11-03 04:00:00-06:00   NaN
Freq: H, dtype: float64

In [4]: pd.Series(index=idx).shift(freq='H')
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-4-19ee418c9aa1> in <module>()
----> 1 pd.Series(index=idx).shift(freq='H')

/Users/afni/homebrew/Cellar/python/2.7.5/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/core/generic.pyc in shift(self, periods, freq, axis, **kwds)
   3290             new_data = self._data.shift(periods=periods, axis=block_axis)
   3291         else:
-> 3292             return self.tshift(periods, freq, **kwds)
   3293 
   3294         return self._constructor(new_data).__finalize__(self)

/Users/afni/homebrew/Cellar/python/2.7.5/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/core/generic.pyc in tshift(self, periods, freq, axis, **kwds)
   3386         else:
   3387             new_data = self._data.copy()
-> 3388             new_data.axes[block_axis] = index.shift(periods, offset)
   3389 
   3390         return self._constructor(new_data).__finalize__(self)

/Users/afni/homebrew/Cellar/python/2.7.5/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/tseries/index.pyc in shift(self, n, freq)
    855         end = self[-1] + n * self.offset
    856         return DatetimeIndex(start=start, end=end, freq=self.offset,
--> 857                              name=self.name, tz=self.tz)
    858 
    859     def repeat(self, repeats, axis=None):

/Users/afni/homebrew/Cellar/python/2.7.5/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/tseries/index.pyc in __new__(cls, data, freq, start, end, periods, copy, name, tz, verify_integrity, normalize, closed, **kwds)
    204             return cls._generate(start, end, periods, name, freq,
    205                                  tz=tz, normalize=normalize, closed=closed,
--> 206                                  infer_dst=infer_dst)
    207 
    208         if not isinstance(data, (np.ndarray, ABCSeries)):

/Users/afni/homebrew/Cellar/python/2.7.5/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pandas/tseries/index.pyc in _generate(cls, start, end, periods, name, offset, tz, normalize, infer_dst, closed)
    369         if tz is not None and inferred_tz is not None:
    370             if not inferred_tz == tz:
--> 371                 raise AssertionError("Inferred time zone not equal to passed "
    372                                      "time zone")
    373 

AssertionError: Inferred time zone not equal to passed time zone
```

Playing with `dateutil` and `pytz` timezones makes me despair of repairing that assertion to be anything sane.  Is it actually needed, or can we just turn it into a warning?

Observations:
- two `pytz.DstTzInfo`s of the same zoneinfo zone that have been `.normalize()`ed to different times may compare different
- a `pytz.*TzInfo` has a `zone` member containing the name of the zoneinfo zone from which it was constructed
- a `dateutil.tzfile` has no public members beyond the `tzinfo` API, which is insufficient to tell whether two time zones are equal
- a `pytz` time zone and a `dateutil` time zone appear to compare different under all circumstances, regardless of whether they represent the same time zone
- the various `dateutil` time zone classes have no common base class within the `dateutil` package
- `pytz` zones constructed from different names for the same zoneinfo zone (e.g. `UTC` and `Etc/UTC`) compare different

Ugh.

from #8817

```
import pandas as pd
import pytz
import datetime

dt = datetime.datetime(2014, 11, 14, 0)
dt_est = pytz.timezone('EST').localize(dt)
s = pd.Series(data=[1], index=[dt_est])

s.shift(0, freq='h')  # 2014-11-14 00:00:00-05:00 (seems okay) 
s.shift(-1, freq='h')  # 2014-11-13 18:00:00-05:00 (expected 2014-11-13 23:00:00)
s.shift(1, freq='h')  # 2014-11-13 20:00:00-05:00 (expected 2014-11-14 01:00:00)

s.shift(-1, freq='s')  # 2014-11-13 18:59:59-05:00 (same with other freq)
```

### LINKED ISSUE #8910: TST/BUG: Improve error handling when parsing iso 8601 strings
xref #8832 

Parsing of datetime strings ending with `Z0`or `Z00`should raise (this is currently only the case with `Z0`).
This is due to the fact that the parser in `np_datetime_strings.c` cannot raise its own errors since in tslib.pyx, function convert_ts_tsobject(), any ValueError is catched and then a fallback external parser (dateutil) is called.

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #8616
- [x] closes #8910
- [x] closes #7825
- [x] closes #8306
- [x] closes #8531
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

In the spirit of your https://github.com/pandas-dev/pandas/pull/21407#issuecomment-396896678 @jreback, cleaned up some old timezone issues (but not necessarily covered by #21407) that appear to have been solved

### CODE REVIEW

**Comment by jreback** [comment]
@mroeschke looks awesome!

so can you either add these issue numbers to the whatsnew for #21407, or if its a 'different' issue (even though it was solved via the same), add an appropriate whatnew e.g. a resampling fixed deserves its own whatsnew (in bug fixes for all of these i think)

**Comment by mroeschke** [comment]
Thanks @jreback. Yeah I don't think #21407 was directly related to any of these older issues, so made separate entries for each.

**Comment by jreback** [comment]
thanks @mroeschke !

File: doc/source/whatsnew/v0.23.2.txt [hunk 1]
```diff
@@ -79,6 +79,11 @@ Bug Fixes
 **Timezones**
 - Bug in :class:`Timestamp` and :class:`DatetimeIndex` where passing a :class:`Timestamp` localized after a DST transition would return a datetime before the DST transition (:issue:`20854`)
 - Bug in comparing :class:`DataFrame`s with tz-aware :class:`DatetimeIndex` columns with a DST transition that raised a ``KeyError`` (:issue:`19970`)
+- Bug in :meth:`DatetimeIndex.shift` where an ``AssertionError`` would raise when shifting across DST (:issue:`8616`)
+- Bug in :class:`Timestamp` constructor where passing an invalid timezone offset designator (``Z``) would not raise a ``ValueError``(:issue:`8910`)
+- Bug in :meth:`Timestamp.replace` where replacing at a DST boundary would retain an incorrect offset (:issue:`7825`)
+- Bug in :meth:`DatetimeIndex.reindex` when reindexing a tz-naive and tz-aware :class:`DatetimeIndex` (:issue:`8306`)
+- Bug in :meth:`DatetimeIndex.resample` when downsampling across a DST boundary (:issue:`8531`)
```
[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/whatsnew/v0.23.2.txt]: see #21407

[the remark this answers is not in the document]
Comment (mroeschke) [reply doc/source/whatsnew/v0.23.2.txt]: These all didn't get fixed by #21407 per se. These issues may have been fixed by other prior PR's in other releases. What should our whatsnew policy be on cleaning up these older issues that have already been fixed?

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply doc/source/whatsnew/v0.23.2.txt]: I don't know if it is needed to list them at all. But if we list them, I would rather add them eg to the 0.24.0 (or 0.23.0) file than to the 0.23.2 file. For those bug fix releases, the list is much shorter and it is often more valuable to actually look which bugs/regressions have been fixed compared to 0.23.0 (in the 0.24.0 or 0.23.0 files the list is huge anyways ...)

**Comment by jreback** [comment]
it’s fine to move some to 0.24 whatsnew
though if they r indeed fixed by #21407 are ok here

**Comment by mroeschke** [comment]
Sounds good. I have another timezones issues PR in the works so I'll keep that in mind

**Pull request merged**